### DISCUSSION TITLE
WiX 4 - bafunctions - header files?

### DISCUSSION THREAD

**Discussion by mtissington** [wixtoolset/issues#7616] [Questions] [answered]
I'm upgrading from Wix 3.x and using Visual Studio 2022.
I have a project called bafunctions and uses IBootstrapperBAFunction

This was defined in the wixtoolset for 3.x and I could reference the various header files etc.

In Wix 4 how do I do this?
Where are the header files?

**Comment by BMurri** [comment, marked as answer]
NuGet package [WixToolset.BalUtil](https://www.nuget.org/packages/WixToolset.BalUtil/).

Also, some [changes were made](https://github.com/wixtoolset/wix/tree/v4.0.1/src/api/burn/balutil/inc) (base type name changes, etc.)

You may want to search for `<ProjectReference Include="WixToolset.BalUtil"` in the wix sources for examples.

**Reply by mtissington** [reply]
Thanks - I found the following [example](https://github.com/wixtoolset/wix/tree/cef14c6055f85e470ff9ce7a33b53e80d1160ba6/src/ext/Bal/Samples/bafunctions)

What package is required for Custom Actions?

**Reply by BMurri** [reply]
native (c++) CAs: `WixToolset.WcaUtil` (which itself depends on `WixToolset.DUtil`)
managed (c#) CAs: `WixToolset.Dtf.CustomAction` (which itself depends on `WixToolset.Dtf.WindowsInstaller`)
